s3-handler/main.go

```diff
@@ -71,7 +79,19 @@ func Handler(request events.S3Event) (Response, error) {
 				}).Info("parser checkpoint")
 			}
 
-			if sampleRate != 1 && rand.Intn(int(sampleRate)) != 0 {
+			sampleRateForLine := sampleRate
+
+			if len(sampleRateRules) > 0 {
```

Comment:
You don't need this check. `for ... range` will do the right thing with an empty or nil slice.